File: anthill_front/lib/src/modules/transactions/domain/dtos/transaction_stats_dto.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../../../shared/utils/date_transfer_format.dart';

part 'transaction_stats_dto.freezed.dart';

part 'transaction_stats_dto.g.dart';

@freezed
class TransactionStatsDto with _$TransactionStatsDto {
  const factory TransactionStatsDto({
    required DateTime fromDate,
    required DateTime toDate,
    required double largestIncome,
    required double averageIncome,
    required double incomesSum,
    required int incomesCount,
    required double expensesSum,
    required int expensesCount,
    @JsonKey(fromJson: _balancesFromJson) required Map<DateTime, List<double>> balances,
  }) = _TransactionStatsDto;

  factory TransactionStatsDto.fromJson(Map<String, Object?> json) =>
      _$TransactionStatsDtoFromJson(json);
}

Map<DateTime, List<double>> _balancesFromJson(Map<String, dynamic> json) {
  return json.map((key, value) {
    final dateKey = deserializeDateQueryParam(key)!;

    return MapEntry(dateKey, value as List<double>);
  });
}

File: anthill_back/src/modules/transactions/data/dtos/transaction.stats.dto.ts
import { ApiProperty } from '@nestjs/swagger';

export class TransactionStatsDto {
  fromDate: Date;
  toDate: Date;
  largestIncome: number;
  averageIncome: number;
  incomesSum: number;
  incomesCount: number;
  expensesCount: number;
  expensesSum: number;

  // a map from a Date string representation (YYYY-MM-DD)
  // to total amount received at that day
  @ApiProperty({
    name: 'balances',
    type: {},
    example: {
      '2024-03-18': 12.3,
      '2024-03-17': 3.4,
    },
  })
  balances: {
    [key: string]: number;
  };
}
